File: lib/collection/vec.cs
module collection.vec

import * from datatype.nat

type vec_t(A: type_t, length: nat_t) {
  null_vec_t(length = zero_t)
  cons_vec_t(n: nat_t, length = succ_t(n), head: A, tail: vec_t(A, n))
}

@infix(++)
vec_append(ante: vec_t(A, m), succ: vec_t(A, n)): vec_t(A, m + n) {
  ante choice {
    null_vec_t => succ
    cons_vec_t => cons_vec_t(ante.head, ante.tail ++ succ)
  }
}

vec_map(f: A -> B, vec: vec_t(A, n)): vec_t(A, n) {
  vec choice {
    null_vec_t => vec
    cons_vec_t => cons_vec_t(f(vec.head), vec_map(f, vec.tail))
  }
}

File: src/main/scala/xieyuheng/cicada/with_logic_variable/prelude/vec.scala
package xieyuheng.cicada.with_logic_variable.prelude

import xieyuheng.cicada.with_logic_variable._
import xieyuheng.cicada.with_logic_variable.expDSL._

object vec extends Module {

  import_all(nat)

  // define_type("vec_t", %(
  //   "A" -> the_type,
  //   "length" -> the("nat_t")),
  //   members = %(
  //     "null_vec_t" -> %(
  //       "length" -> "zero_t"),
  //     "cons_vec_t" -> %(
  //       "n" -> the("nat_t"),
  //       "length" -> ("succ_t" ap %("prev" -> "n")),
  //       "head" -> the("A"),
  //       "tail" -> the("vec_t" ap %("A" -> "A", "length" -> "n")))))


  define_type("vec_t", %(
    "A" :: Type(),
    "length" :: "nat_t"),
    members = %(
      "null_vec_t" -> %(
        "length" := "zero_t"),
      "cons_vec_t" -> %(
        "n" :: "nat_t",
        "length" := ("succ_t" ap %("prev" -> "n")),
        "head" :: "A",
        "tail" :: ("vec_t" ap %("A" -> "A", "length" -> "n")))))

  // TODO
  // vec_map
  // vec_append
}
